### PR TITLE
Adding test coverage for host builder properties

### LINKED ISSUE #36197: 🐞 `WebApplication.CreateBuilder(args).Host.Properties` cannot share context value.
Use the latest `WebApplication.CreateBuilder(args)` to create the object `builder` and set it's `Properties` value through the `Host` property. 

For example, add a `data` element, but the value just set cannot be obtained in the `ConfigureContainer` and `UseServiceProviderFactory` methods. It will always be displayed here as the initial value.

However, in the past version (`.NET5`), it was normal through the `Host.CreateDefaultBuilder(args)`.

See the detailed code and screenshot below.

```cs
var builder = WebApplication.CreateBuilder(args);

// step1: add some share data
builder.Host.Properties.Add("data", "aspnetcore");

builder.Host.ConfigureContainer<IServiceCollection>((hostContext, services) =>
{
    // step2: the data is lost
    var data = hostContext.Properties; // the data is lost
});
```

image

```cs
var builder = Host.CreateDefaultBuilder(args);

// step1: add some share data
builder.Properties.Add("data", "aspnetcore");

builder.ConfigureContainer<IServiceCollection>((hostContext, services) =>
{
    // step2: the data is normal
    var data = hostContext.Properties; // the data is normal
});
```

image

### PR DESCRIPTION
**Pull request by davidfowl**
- Adds test coverage to make sure properties are propagated through callbacks.

Closes #36197

### CODE REVIEW

**Review by captainsafia — APPROVED** [review]
LGTM.

Will we get any value out of adding a scenario to validate that the `build.Properties.Add("foo", "bar")` scenario works?

**Comment by davidfowl** [comment]
> Will we get any value out of adding a scenario to validate that the build.Properties.Add("foo", "bar") scenario works?

I don't think so, we're not using a custom dictionary or anything. Add vs set in this case is equivalent.

**Comment by captainsafia** [comment]
> I don't think so, we're not using a custom dictionary or anything. Add vs set in this case is equivalent.

I was moreso referring to testing `builder.Properties` vs `builder.Host.Properties`.

**Comment by davidfowl** [comment]
There's no `WebApplicatioBuilder.Properties`.

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 1]
```diff
@@ -1373,6 +1373,36 @@ public async Task NoExceptionAreThrownForBadRequestsInProduction()
...
+                Assert.Equal("world", context.Properties["hello"]);
+            });
+
+            using var app = builder.Build();
+
+            // Make sure all of the callbacks ran
+            Assert.Equal(0b00000111, callbacks);
+        }
+
         public void HostConfigurationNotAffectedByConfiguration()
```
Comment (captainsafia) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:1406]:
```suggestion
        [Fact]
        public void HostConfigurationNotAffectedByConfiguration()
```

**Comment by davidfowl** [comment]
And you better not have a merge conflict 😄

**Pull request merged**